// @trezor/transport/lib/transports/nodeusb.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NodeUsbTransport = void 0;
const usb_1 = require("usb");
const abstractApi_1 = require("./abstractApi");
const client_1 = require("../sessions/client");
const background_1 = require("../sessions/background");
const usb_2 = require("../api/usb");
class NodeUsbTransport extends abstractApi_1.AbstractApiTransport {
    constructor(params) {
        const { messages, logger, debugLink } = params;
        const sessionsBackground = new background_1.SessionsBackground();
        const sessionsClient = new client_1.SessionsClient({
            requestFn: args => sessionsBackground.handleMessage(args),
            registerBackgroundCallbacks: () => { },
        });
        sessionsBackground.on('descriptors', descriptors => {
            sessionsClient.emit('descriptors', descriptors);
        });
        super({
            messages,
            api: new usb_2.UsbApi({
                usbInterface: new usb_1.WebUSB({
                    allowAllDevices: true,
                }),
                logger,
                debugLink,
            }),
            sessionsClient,
        });
        this.name = 'NodeUsbTransport';
        this.sessionsBackground = sessionsBackground;
    }
    listen() {
        this.api.listen();
        return super.listen();
    }
    stop() {
        super.stop();
        this.sessionsBackground.dispose();
    }
}
exports.NodeUsbTransport = NodeUsbTransport;
