// @trezor/transport/lib/transports/webusb.browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebUsbTransport = void 0;
const abstractApi_1 = require("./abstractApi");
const client_1 = require("../sessions/client");
const usb_1 = require("../api/usb");
const background_browser_1 = require("../sessions/background-browser");
class WebUsbTransport extends abstractApi_1.AbstractApiTransport {
    constructor(params) {
        const { messages, logger } = params;
        const { requestFn, registerBackgroundCallbacks } = (0, background_browser_1.initBackgroundInBrowser)();
        super({
            messages,
            api: new usb_1.UsbApi({
                usbInterface: navigator.usb,
                logger,
            }),
            logger,
            sessionsClient: new client_1.SessionsClient({
                requestFn,
                registerBackgroundCallbacks,
            }),
        });
        this.name = 'WebUsbTransport';
    }
    listen() {
        this.api.listen();
        return super.listen();
    }
}
exports.WebUsbTransport = WebUsbTransport;
